pkg/appmgmt/appmgmt_recovery.go

```diff
@@ -71,33 +71,47 @@ func (svc *AppManagementService) recoverApps() (map[string]interfaces.ManagedApp
 	return recoveringApps, nil
 }
 
-func (svc *AppManagementService) waitForAppRecovery(
-	recoveringApps map[string]interfaces.ManagedApp, maxTimeout time.Duration) error {
-	if len(recoveringApps) > 0 {
-		log.Logger().Info("wait for app recovery",
-			zap.Int("appToRecover", len(recoveringApps)))
-		// check app states periodically, ensure all apps exit from recovering state
-		if err := utils.WaitForCondition(func() bool {
-			for _, app := range recoveringApps {
-				log.Logger().Debug("appInfo",
-					zap.String("appId", app.GetApplicationID()),
-					zap.String("state", app.GetApplicationState()))
-				if app.GetApplicationState() == cache.ApplicationStates().Accepted {
-					delete(recoveringApps, app.GetApplicationID())
-				}
-			}
-
-			if len(recoveringApps) == 0 {
-				log.Logger().Info("app recovery is successful")
-				return true
-			}
-
+func (svc *AppManagementService) waitForAppRecovery(recoveringApps map[string]interfaces.ManagedApp) bool {
+	svc.cancelRecovery.Store(false) // reset cancellation token
+	recoveryStartTime := time.Now()
+	counter := 0
+	for {
+		// check for cancellation token
+		if svc.cancelRecovery.Load() {
+			log.Logger().Info("Waiting for recovery canceled.")
+			svc.cancelRecovery.Store(false)
```

Comment:
why reset? if recovery is cancelled we should never re-run it and a new recovery sets it to false on line 75

Reply:
This is for test code that runs this multiple times to check negative constraints.

Reply:
Specifically, TestAppStatesDuringRecovery.